contracts/VolumeImbalanceRecorder.sol

```diff
@@ -0,0 +1,152 @@
+pragma solidity ^0.4.18;
+
+import "./ERC20Interface.sol";
+import "./PermissionGroups.sol";
+
+interface VolumeImbalanceRecorderInterface {
+  function getImbalance( ERC20 token, uint priceUpdateBlock, uint currentBlock)
+    public view returns(int totalImbalance, int currentBlockImbalance);
+}
+
+contract VolumeImbalanceRecorder is PermissionGroups, VolumeImbalanceRecorderInterface {
+
+  uint public constant SLIDING_WINDOW_SIZE = 10;
+
+  struct TokenControlInfo {
+    uint minimalRecordResolution; // can be roughly 1 cent
+    uint maxPerBlockImbalance; // in twei resolution
+    uint maxTotalImbalance; // max total imbalance (without price updates)
+                            // before halting trade
+  }
+
+  mapping(address => TokenControlInfo) tokenControlInfo;
+
+  struct TokenImbalanceData {
+    int64  lastBlockBuyUnitsImbalance;
+    uint64 lastBlock;
+
+    int64  totalBuyUnitsImbalance;
+    uint64 lastPriceUpdateBlock;
+  }
+
+  mapping(address => mapping(uint=>TokenImbalanceData)) tokenImbalanceData;
+
+  event SetTokenControlInfo( ERC20 token,
+                             uint minimalRecordResolution,
+                             uint maxPerBlockImbalance,
+                             uint maxTotalImbalance );
+
+  function setTokenControlInfo( ERC20 token,
+                                uint minimalRecordResolution,
+                                uint maxPerBlockImbalance,
+                                uint maxTotalImbalance ) public onlyAdmin {
+    tokenControlInfo[token] = TokenControlInfo(
+                                minimalRecordResolution,
+                                maxPerBlockImbalance,
+                                maxTotalImbalance);
+    SetTokenControlInfo( token,
+                         minimalRecordResolution,
+                         maxPerBlockImbalance,
+                         maxTotalImbalance );
+  }
+
+  function getTokenControlInfo( ERC20 token ) public view returns(uint,uint,uint) {
+    return (tokenControlInfo[token].minimalRecordResolution,
+            tokenControlInfo[token].maxPerBlockImbalance,
+            tokenControlInfo[token].maxTotalImbalance);
+  }
+
+  function getImbalanceInRange( ERC20 token, uint startBlock, uint endBlock ) public view returns(int buyImbalance ) {
+    // check the imbalance in the sliding window
+    require(startBlock <= endBlock);
+
+    buyImbalance = 0;
+
+    for(uint windowInd = 0 ; windowInd < SLIDING_WINDOW_SIZE ; windowInd++ ) {
+      TokenImbalanceData memory perBlockData = tokenImbalanceData[token][windowInd];
+      if( perBlockData.lastBlock <= endBlock && perBlockData.lastBlock >= startBlock ) {
+        buyImbalance += int(perBlockData.lastBlockBuyUnitsImbalance);
+      }
```

Comment:
whenever range is larger then sliding window size, the results could include "old blocks" data if no block was used in the middle. but if it was the "old block" data will be erased... so in a way this is inconsistent.

Reply:
it is inconsistent, and only best effort. it is possible to add some verification, but I am not sure if needed. As long as it fits into the sliding window it should be ok.